kotlinx-coroutines-core/jvm/src/internal/ResizableAtomicArray.kt

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2016-2022 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.internal
+
+import java.util.concurrent.atomic.*
+
+/**
+ * Atomic array with lock-free reads and synchronized modifications. It logically has an unbounded size,
+ * is implicitly filled with nulls, and is resized on updates as needed to grow.
+ */
+internal class ResizableAtomicArray<T>(initialLength: Int) {
+    @Volatile
+    private var array = AtomicReferenceArray<T>(initialLength)
+
+    // for debug output
+    public fun currentLength(): Int = array.length()
+
+    public operator fun get(index: Int): T? {
+        val array = this.array // volatile read
+        return if (index < array.length()) array[index] else null
+    }
+
+    // Must not be called concurrently, e.g. always use synchronized(this) to call this function
+    fun setSynchronized(index: Int, value: T?) {
+        val curArray = this.array
+        val curLen = curArray.length()
+        if (index < curLen) {
+            curArray[index] = value
+        } else {
+            val newArray = AtomicReferenceArray<T>((index + 1).coerceAtLeast(2 * curLen))
+            for (i in 0 until curLen) newArray[i] = curArray[i]
+            newArray[index] = value
+            array = newArray // copy done
```

Comment:
```
class test<E> (initialLength: Int = 2) {
    
    private var array: AtomicReferenceArray<E?> = AtomicReferenceArray(initialLength)
    private val size: AtomicInteger = AtomicInteger(initialLength)

    fun get(idx: Int): E? = idx.coerceAtMost(size.get()).let { array.get(it) }

    fun set(idx: Int, newValue: E?) {
        size.get().takeIf { it < idx }?.let { s ->
            arrayListOf(
                0.rangeTo(s).asSequence().map { array.get(it) }.toList()
            ).toArray().copyOf(
                idx.inc().coerceAtLeast(s.times(2))
            ).let { newArray ->
                AtomicReferenceArray(newArray as Array<out E?>).let {
                    array = it
                    size.set(newArray.size)
                }
            }
        }
        array.set(idx, newValue)
    }
}

```
synchronized and Volatile do not need

Reply:
I'm not sure what your comments mean, but the implementation you've provided is clearly non-linearizable (however, our current testing infrastructure is unlikely to catch it).